Fix file sync issue when background upload finishes but doesn't update state in main app

michalrentka opened the pull request:
Background uploads can sometime pass and file is successfully uploaded to backend, but main app is not notified and local state is not updated, so when user opens main app it tries to upload the same file during sync again and receives 412 `If-None-Match: * set but file exists`. In this case we re-download the remote file version to be sure both are the same and mark file as uploaded.

dstillman commented on Zotero/Controllers/Sync/SyncController.swift:
```diff
@@ -1747,12 +1771,23 @@ final class SyncController: SynchronizationController {
                 return true
             }
 
+            if response == "If-None-Match: * set but file exists" {
```
Don't want to test against this — regular English strings (i.e., not JSON with a specific structure) in the response body are just for debugging and aren't guaranteed to be stable. The relevant conditions here would be 1) a `412` for a file upload authorization request and 2) `Last-Modified-Version` of the response > the local attachment version. And then there are some other situations that the desktop app handles if (2) isn't true:

https://github.com/zotero/zotero/blob/6106e379c96a26897ead32a300190e2327a11f3e/chrome/content/zotero/xpcom/storage/zfs.js#L552-L620

Also doesn't seem right that this is under something called `libraryConflict`. This would specifically be for a file upload authorization request. You can see the other 4xx response codes handled by the desktop for a file upload authorization request in the function linked above. I would expect this to be handled in the same place.

michalrentka replied:
@dstillman I'm not sure I understand fully. In case of 412 on authorization it first checks response header version, compares to local item version and if remote version is newer, it re-downloads the file and updates item. 

But if remote `version <= item.version` then it fetches the same item from backend and checks versions again? Wouldn't `Last-Modified-Version` from authorization response be the same as `version` of item fetched directly from backend?

Then if remote `version <= item.version` it checks md5 hashes of remote and local files, if they match it just updates the local file as synced. If they don't match, it marks a file conflict, which we probably don't support on iOS yet. So I guess I'll just abort sync in that case?

Also for 413 I guess I'll just show an error with usage and quota values?

dstillman replied:
Yeah, sorry, that's really confusing ("downloading" is referring to the attachment item, not the file), possibly evolved incorrectly over the years, and probably isn't relevant in any case as long as there's no local file editing.

It looks like there are three possible causes of 412 for a file upload authorization request currently:

1) If-Match being set to a hash but the file not existing
2) The ETag not matching the current version of the file
3) If-None-Match: * being set but the file existing

(1) won't ever happen from iOS, I don't think.

(2) and (3) can probably be simplified to force-downloading the file and marking as uploaded, yes, at least as long as we don't have local file editing.

michalrentka commented:
@dstillman I updated `SyncController` to handle upload authorization failures separately as requested: https://github.com/zotero/zotero-ios/blob/dfa41cb6de595a5808a9a90bb4aa7142203d915e/Zotero/Controllers/Sync/SyncController.swift#L1646-L1700. 

And I implemented the logic to revert file uploads in case of 403 on file authorization: https://github.com/zotero/zotero-ios/blob/dfa41cb6de595a5808a9a90bb4aa7142203d915e/Zotero/Controllers/Sync/SyncActions/RevertLibraryFilesSyncAction.swift.

There's also a bit of refactoring and simplification of `SyncScheduler` and cleanups due to us adding SwiftLint to code base.

mvasilak commented on Zotero/Controllers/Sync/SyncActions/UploadFixSyncAction.swift:
```diff
@@ -0,0 +1,168 @@
+                        case .remoteMissing:
+                            DDLogError("UploadFixSyncAction: attachment missing remotely")
+                            subscriber(.failure(Error.attachmentMissingRemotely))
+                            return Disposables.create()
+
+                        case .local, .localAndChangedRemotely, .remote: break
+                        }
+
+                        // Create new attachment model with updated location so that `AttachmentDownloader` doesn't ignore it
+                        let newAttachment = Attachment(type: .file(filename: filename, contentType: contentType, location: .remote, linkType: linkType),
+                                                       title: attachment.title, key: attachment.key, libraryId: attachment.libraryId)
+                        subscriber(.success(newAttachment))
```
This seems a bit more clear, since switch is exhaustive.

```suggestion

                        case .local, .localAndChangedRemotely, .remote:
                            // Create new attachment model with updated location so that `AttachmentDownloader` doesn't ignore it
                            let newAttachment = Attachment(type: .file(filename: filename, contentType: contentType, location: .remote, linkType: linkType),
                                                           title: attachment.title, key: attachment.key, libraryId: attachment.libraryId)
                            subscriber(.success(newAttachment))
                        }
```

michalrentka replied:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0

mvasilak commented on Zotero/Controllers/Sync/SyncActions/UploadFixSyncAction.swift:
```diff
@@ -0,0 +1,168 @@
+
+                case .ready:
+                    self.finishDownload?(.success(()))
+                    self.finishDownload = nil
+                case .progress, .cancelled: break
```
```suggestion
                    self.finishDownload = nil
                    
                case .progress, .cancelled: 
                    break
```

michalrentka replied:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0

mvasilak commented on Zotero/Controllers/Sync/SyncScheduler.swift:
```diff
@@ -8,180 +8,219 @@
-private typealias SchedulerAction = (syncType: SyncController.SyncType, librarySyncType: SyncController.LibrarySyncType)
-
+/// Controller that schedules synchronisation of local and remote data. All syncs are requested through this controller and it decides if/when next sync should be started.
+/// If a sync fails this controller can also retry it with to resolve issues.
```
Something like:
```suggestion
/// If a sync fails this controller can also retry it with increasing intervals to resolve issues.
```

michalrentka replied:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0

mvasilak commented on Zotero/Controllers/Sync/SyncScheduler.swift:
```diff
@@ -8,180 +8,219 @@
+        self.enqueueAndStart(sync: Sync(type: type, libraries: libraries))
     }
 
+    /// Requests a sync based on websocked reported changes in given library.
```
```suggestion
    /// Requests a sync based on websocket reported changes in given library.
```

michalrentka replied:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0

mvasilak commented on Zotero/Controllers/Sync/SyncScheduler.swift:
```diff
@@ -8,180 +8,219 @@
-            self.nextAction = action
+    /// Start next sync in queue. In case of normal/external syncs wait at least `syncTimeout` between syncs. In case of retry syncs wait for specified delay based on current retry attempt count.
+    private func startNextSync() {
+        // Ignore this if there is a sync in progress. Otherwise pick next sync in queue.
+        guard self.syncInProgress == nil, let nextSync = self.syncQueue.first else {
+            if self.syncInProgress == nil && self.syncQueue.isEmpty {
+                // Report finished queue
+                self.inProgress.accept(false)
+            }
```
This seems a bit more clear:
```suggestion
        // Ignore this if there is a sync in progress.
        guard self.syncInProgress == nil else { return }
        // Otherwise pick next sync in queue.
        guard let nextSync = self.syncQueue.first else {
            // Report finished queue
            self.inProgress.accept(false)
```

michalrentka replied:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0

michalrentka commented on Zotero/Controllers/Sync/SyncActions/RevertLibraryFilesSyncAction.swift:
```diff
@@ -0,0 +1,101 @@
+            } catch let error {
+                DDLogWarn("RevertLibraryFilesSyncAction: can't rename file - \(error)")
+                // If it can't be moved, at least delete the old one. It'll have to be re-downloaded anyway.
+                try? self.fileStorage.remove(oldFile)
```
I mean, there's always a chance :). If something broke previously during sync and the name didn't change properly... So removing the new file too won't hurt.

michalrentka replied:
Check 49628051723bedbfb92ec2c57f4cc9f7282ec56b

dstillman commented:
Do we have tests for these things?

michalrentka commented:
> Do we have tests for these things?

They are in line after it's approved. I fixed existing tests in `testfixes` branch. I'll add new test for this file reverting mechanism (do we have any on desktop app?) and then we can gradually add more as needed. I'll create a new ticket for me and @mvasilak with a (mostly) prioritized list of tests so that we can increase our test coverage together.
